### PR TITLE
Fixed issue #19481: After composer installation sample data can't be installed from command line

### LINKED ISSUE #19481: Magento 2.3.0: After composer installation sampledata can't be installed from command line
### Preconditions
1. Magento 2.3.0
2. Install Magento with composer
3. Remove the version number from the composer.json file in the root of the project.

### Steps to reproduce
1. Install Magento with composer using the following command `composer create-project --repository=https://repo.magento.com/ magento/project-community-edition <install-directory-name>`
2. After successful installation and setup run the following command to install the sample data: `bin/magento sampledata:deploy`

### Expected result
1. Sample data deploys to the freshly installed Magento environment.

### Actual result
1. Command throws the following error: 

```
bin/magento sampledata:deploy
Git installations must deploy sample data from GitHub; see https://devdocs.magento.com/guides/v2.3/install-gde/install/sample-data-after-clone.html for more information.
```

image

### PR DESCRIPTION
**Pull request by andrewbess**
### Description (*)

This PR fixes the problem when sample data aren't deploying if the field "version" has removed from composer.json.

### Fixed Issues (if relevant)

1. Fixes magento/magento2#19481: Magento 2.3.0: After composer installation sampledata can't be installed from command line

### Manual testing scenarios (*)

1. Install Magento with the composer;
2. Remove the version number from the composer.json file in the root of the project;
3. Try to deploy sample data using `bin/magento sampledata:deploy`

### Actual result
1. Command throws the following error:
```
bin/magento sampledata:deploy
Git installations must deploy sample data from GitHub; see https://devdocs.magento.com/guides/v2.3/install-gde/install/sample-data-after-clone.html for more information.
```
console-output-before-fix

2. The sample data haven't deployed.

### Result after fix

1. Command throws the following message:
```
bin/magento sampledata:deploy
We don't recommend to remove the "version" field from your composer.json; see https://getcomposer.org/doc/02-libraries.md#library-versioning for more information.
The field "version" has been restored.
```
console-output-after-fix

2. The field "version" has been restored in `composer.json`;

3. The sample data have been deployed successfully.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/SampleData/Console/Command/SampleDataDeployCommand.php [hunk 1]
```diff
@@ -79,15 +86,42 @@ protected function configure()
...
+     * @throws LocalizedException
      */
     protected function execute(InputInterface $input, OutputInterface $output)
     {
-        $rootJson = json_decode($this->filesystem->getDirectoryRead(DirectoryList::ROOT)->readFile("composer.json"));
+        $rootJson = json_decode(
+            $this->filesystem->getDirectoryRead(
+                DirectoryList::ROOT
+            )->readFile("composer.json")
+        );
```
Comment (lbajsarowicz) [inline app/code/Magento/SampleData/Console/Command/SampleDataDeployCommand.php]: Could you use Json Serializer from Magento package?

File: app/code/Magento/SampleData/Console/Command/SampleDataDeployCommand.php [hunk 2]
```diff
@@ -79,15 +86,42 @@ protected function configure()
...
         if (!isset($rootJson->version)) {
-            // @codingStandardsIgnoreLine
-            $output->writeln('<info>' . 'Git installations must deploy sample data from GitHub; see https://devdocs.magento.com/guides/v2.3/install-gde/install/sample-data-after-clone.html for more information.' . '</info>');
-            return;
+            $magentoProductPackage = array_filter(
+                (array) $rootJson->require,
```
Comment (lbajsarowicz) [inline app/code/Magento/SampleData/Console/Command/SampleDataDeployCommand.php]: I'm not sure if you linted your code properly (let's wait for Static tests)

**Review by lbajsarowicz — APPROVED** [review]
There's just a minor change + code lint that is going to be detected by Static Tests.
Logic and the approach is - in my opinion - correct. Especially taking into consideration additional notice about missing `version` node.

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-7219](https://jira.corp.magento.com/browse/ENGCOM-7219) has been created to process this Pull Request
:eight_spoked_asterisk: @lbajsarowicz, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

File: app/code/Magento/SampleData/Console/Command/SampleDataDeployCommand.php [hunk 3]
```diff
@@ -50,22 +53,30 @@ class SampleDataDeployCommand extends Command
...
      * @param ApplicationFactory $applicationFactory
+     * @param SerializerInterface $serializer
      */
     public function __construct(
         Filesystem $filesystem,
         Dependency $sampleDataDependency,
         ArrayInputFactory $arrayInputFactory,
-        ApplicationFactory $applicationFactory
+        ApplicationFactory $applicationFactory,
+        SerializerInterface $serializer
```
Comment (lbajsarowicz) [inline app/code/Magento/SampleData/Console/Command/SampleDataDeployCommand.php]: I think Magento is not the guy you'd like to trust when it comes to Serializer interface :-P
Use the one `\Magento\Framework\Serialize\Serializer\Json`

File: app/code/Magento/SampleData/Test/Unit/Console/Command/AbstractSampleDataCommandTest.php [hunk 4]
```diff
@@ -181,4 +181,19 @@ protected function setupMocks(
...
+     * @return array
+     */
+    protected function expectedComposerArgumentsCommandConfig(): array
+    {
+        return [
+            'command' => 'config',
+            'setting-key' => 'version',
+            'setting-value' => ['0.0.1'],
+            '--quiet' => 1
+        ];
```
Comment (lbajsarowicz) [inline app/code/Magento/SampleData/Test/Unit/Console/Command/AbstractSampleDataCommandTest.php]: That can be just `private const STUB_EXPECTED_COMPOSER_CONFIG`

**Review by lbajsarowicz — APPROVED** [review]
The only issue I see is the serializer interface. That should be explicit `JSON`, as Magento is going to - sooner or later - replace the serializer preference.

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-7219](https://jira.corp.magento.com/browse/ENGCOM-7219) has been created to process this Pull Request
:eight_spoked_asterisk: @lbajsarowicz, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

File: app/code/Magento/SampleData/Test/Unit/Console/Command/AbstractSampleDataCommandTest.php (hunk 4 above)
Comment (andrewbess) [reply app/code/Magento/SampleData/Test/Unit/Console/Command/AbstractSampleDataCommandTest.php]: Hello @lbajsarowicz 
Thank you for your review.
I fixed it. Please check.
Thank you in advance.

File: app/code/Magento/SampleData/Console/Command/SampleDataDeployCommand.php (hunk 3 above)
Comment (andrewbess) [reply app/code/Magento/SampleData/Console/Command/SampleDataDeployCommand.php]: Hello @lbajsarowicz 
Thank you for your review.
I changed SerializerInterface to Serializer\Json. Please check.
Thank you in advance.

File: app/code/Magento/SampleData/Console/Command/SampleDataDeployCommand.php (hunk 1 above)
Comment (andrewbess) [reply app/code/Magento/SampleData/Console/Command/SampleDataDeployCommand.php]: Hello @lbajsarowicz 
Thank you for your review.
I fixed it. Please check.
Thank you in advance.

File: app/code/Magento/SampleData/Console/Command/SampleDataDeployCommand.php [hunk 5]
```diff
@@ -137,30 +196,49 @@ private function createAuthFile()
...
     /**
+     * Updates PHP memory limit
+     *
+     * @throws InvalidArgumentException
      * @return void
      */
     private function updateMemoryLimit()
     {
         if (function_exists('ini_set')) {
-            @ini_set('display_errors', 1);
```
Comment (andrewbess) [inline app/code/Magento/SampleData/Console/Command/SampleDataDeployCommand.php:155]: Hello @lbajsarowicz 
I changed `@ini_set` to `ini_alter` (that's the alias of ini_set) and added throwing an exception when ini_set returns an error.
Please check it.
Thank you in advance.

Comment (lenaorobei) [inline app/code/Magento/SampleData/Console/Command/SampleDataDeployCommand.php:155]: Please use `// phpcs:ignore Magento2.Functions.DiscouragedFunction` for the legacy code. `ini_alter` should be added to the discouraged list as well https://github.com/magento/magento-coding-standard/blob/develop/Magento2/Sniffs/Functions/DiscouragedFunctionSniff.php

Comment (andrewbess) [reply app/code/Magento/SampleData/Console/Command/SampleDataDeployCommand.php:155]: Hello @lenaorobei 
Thank you for your remark.
I have reverted ini_alter to ini_set (without `@`).
Also, today I am going to add all alias to discouraged functions

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Before:** :heavy_multiplication_x: Command throws the following error: Git installations must deploy sample data from GitHub;

Screenshot from 2020-04-16 14-18-43


**After:** :heavy_check_mark: The field "version" has been restored in composer.json and the sample data have been deployed successfully

Screenshot from 2020-04-16 15-18-50

**Review by slavvka — CHANGES_REQUESTED** [review]
@andrewbess please resolve the conflicts and re-run tests after upgrade to PHP 7.4

**Comment by engcom-Echo** [comment]
@magento run all tests

**Review by slavvka — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @slavvka, thank you for the review.
[ENGCOM-7219](https://jira.corp.magento.com/browse/ENGCOM-7219) has been created to process this Pull Request

**Comment by slavvka** [comment]
@magento run all tests

**Review by slavvka — CHANGES_REQUESTED** [review]
My mistake

**Review by slavvka — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @slavvka, thank you for the review.
[ENGCOM-7219](https://jira.corp.magento.com/browse/ENGCOM-7219) has been created to process this Pull Request

**Pull request merged**